tracing/grpc.go

```diff
@@ -0,0 +1,210 @@
+	withPayload := s.options.spanWithPayload != nil && s.options.spanWithPayload(ctx, rs)
+
+	span := trace.FromContext(ctx)
+
+	if withPayload {
+		switch rs := rs.(type) {
+		case *stats.End:
+			attrs := []trace.Attribute{trace.StringAttribute(ResponseErrorKey, rs.Error.Error())}
+			span.Annotate(attrs, "Response error")
+		}
+	}
+
+	s.ServerHandler.HandleRPC(ctx, rs)
+
+	if withHeaders {
+		switch rs := rs.(type) {
+		case *stats.InHeader:
+			attrs := metadataToAttributes(rs.Header, RequestHeaderAnnotationPrefix, s.options.metadataMatcher)
+			span.AddAttributes(attrs...)
+		case *stats.InTrailer:
+			attrs := metadataToAttributes(rs.Trailer, RequestTrailerAnnotationPrefix, s.options.metadataMatcher)
+			span.AddAttributes(attrs...)
+		case *stats.OutHeader:
+			attrs := metadataToAttributes(rs.Header, ResponsePayloadAnnotationKey, s.options.metadataMatcher)
+			span.AddAttributes(attrs...)
+		case *stats.OutTrailer:
+			attrs := metadataToAttributes(rs.Trailer, ResponseTrailerAnnotationPrefix, s.options.metadataMatcher)
+			span.AddAttributes(attrs...)
+		}
+	}
+
+	if withPayload {
+		switch rs := rs.(type) {
+		case *stats.InPayload:
+			attrs, truncated, err := payloadToAttributes(RequestPayloadAnnotationKey, rs.Payload, s.options.maxPayloadSize)
+			if err != nil {
+				ctxlogrus.Extract(ctx).Errorln("unable to marshal response, err - ", err)
+				return
+			}
+
+			if truncated {
+				markSpanTruncated(span)
+			}
+
+			span.Annotate(attrs, "Request payload")
+		case *stats.OutPayload:
+			attrs, truncated, err := payloadToAttributes(ResponsePayloadAnnotationKey, rs.Payload, s.options.maxPayloadSize)
+			if err != nil {
+				ctxlogrus.Extract(ctx).Errorln("unable to marshal response, err - ", err)
+				return
+			}
+
+			if truncated {
+				markSpanTruncated(span)
+			}
+
+			span.Annotate(attrs, "Response payload")
+		}
+	}
+
+}
+
+func metadataToAttributes(md metadata.MD, prefix string, marcher metadataMatcher) []trace.Attribute {
+	attrs := make([]trace.Attribute, 0, len(md))
+	for k, vals := range md {
+		key, ok := marcher(k)
+		if !ok {
+			continue
+		}
+
+		key, valsStr := fmt.Sprint(prefix, key), strings.Join(vals, ", ")
+		if _, ok := sensitiveMetadata[k]; ok {
+			valsStr = obfuscate(valsStr)
+		}
+
+		//Check that key and value is a valid utf-8 encoded strings, in case they not span will be omitted from span
+		if !utf8.ValidString(key) || !utf8.ValidString(valsStr) {
+			continue
+		}
+
+		attrs = append(attrs, trace.StringAttribute(key, valsStr))
+	}
+
+	return attrs
+}
+
+func payloadToAttributes(key string, value interface{}, limit int) ([]trace.Attribute, bool, error) {
+	bytes, err := json.Marshal(value)
+	if err != nil {
+		return nil, false, err
+	}
+
+	bytes, truncated := truncatePayload(bytes, limit)
+	attrs := []trace.Attribute{trace.StringAttribute(key, string(bytes))}
+	return attrs, truncated, nil
+}
+
+func obfuscate(x string) string {
+	countChars := int(float64(len(x)) * 0.20)
```

Comment:
This should probably be a constant.  We should consider making this configurable in the future.

Reply:
The 0.20 value